IKVM.Reflection/Universe.cs

```diff
@@ -145,44 +149,13 @@ public enum UniverseOptions
     public sealed class Universe : IDisposable
     {
 
-#if NETCOREAPP3_1
+#if NETCOREAPP3_1_OR_GREATER
 
-        public static readonly string CoreLibName = "netstandard";
-
-        public static string ReferenceAssembliesDirectory
-        {
-            get
-            {
-                return BuildRefDirFrom(System.Runtime.InteropServices.RuntimeEnvironment.GetRuntimeDirectory());
-            }
-        }
-
-        private static string BuildRefDirFrom(string runtimeDir)
-        {
-            // transform a thing like
-            // C:\Program Files\dotnet\shared\Microsoft.NETCore.App\3.1.7
-            // to
-            // C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\3.1.0\ref\netcoreapp3.1
-
-            var parts = runtimeDir.Split(Path.DirectorySeparatorChar);
-            var n = string.IsNullOrEmpty(parts[parts.Length - 1]) ? parts.Length - 2 : parts.Length - 1;
-            var versionDir = parts[n--];
-            var frameworkDir = parts[n--];
-            var newParts = new string[n + 5];
-            Array.Copy(parts, newParts, n);
-            var suffixParts = new string[] { "packs", frameworkDir + ".Ref", "3.1.0", "ref", "netcoreapp3.1" };
-            Array.Copy(suffixParts, 0, newParts, n, suffixParts.Length);
-            var dir = Path.Combine(newParts);
-            if (!Directory.Exists(dir))
-            {
-                throw new FileNotFoundException("Reference assemblies directory: " + dir);
-            }
-            return dir;
-        }
+        public static readonly string StdLibName = "netstandard";
```

Comment:
I thought that the `mscorlib` functionality was moved into an assembly named [`System.Private.CoreLib`](https://github.com/dotnet/runtime/blob/v5.0.0/src/libraries/System.Private.CoreLib/src/System/Int32.cs) (or at least that seems to be the case since .NET 5.0). This seems to be backed up here: https://github.com/dotnet/roslyn/issues/16211#issuecomment-270462911. Although, later they [mention](https://github.com/dotnet/roslyn/issues/16211#issuecomment-373084209) that it is 

> `System.Runtime` at design time and `System.Private.CoreLib` at runtime, but the latter is an implementation detail


Why is `netstandard` being used here? Could this be an artifact from the early attempts at targeting `netstandard2.0` and later `netstandard2.1`? And could it be the reason for #74?

I am not saying that this is wrong per se, but I am wondering if there is any documentation indicating `netstandard` is what we should be looking for.

Reply:
Because there are type forwards in it, and the code path doesn't differentiate between the static and dynamic compilers. It is not optimal.

Reply:
They mention [here](https://github.com/dotnet/roslyn/issues/16211#issuecomment-373085326):

> You have to pass in all assemblies that are involved in the compilation. Type forwarders are followed but you need to specify both the assembly that defines the forwarder and the assembly that the forwarder forwards to.

Is this being accounted for?

Reply:
If the user passes them in.

Reply:
That said, this PR isn't a complete overhaul of the way assemblies are located and reflected on. It's a minor fix that removes the hard coded attempt at discovering a search path, and unifies two different discovery paths. The rest of the code is still doing what it did before: expecting a core library with the core types like Object, String, etc.

It would be better, in my view, if IKVM discovered these by actually searching the assemblies.

Reply:
That explains why #74 is failing (and why I am getting the same error when trying to convert log4j). However, since we are not actually shipping `netstandard.dll` with IKVM tools (it isn't in the .NET Core 3.1 reference assemblies directory), what is the procedure we should give users to provide a path to it for `ikvmc`?

On my system, I have several copies, but the [advice here](https://stackoverflow.com/a/45886961) is to use `C:\Program Files\dotnet\sdk\2.0.3\Microsoft\Microsoft.NET.Build.Extensions\net461\lib\netstandard.dll`. It sounds like you need Visual Studio to be installed in order for this path to exist, though. And it is a little unsettling that it is in a `net461` folder.

It looks like there are [a couple of NuGet packages](https://github.com/dotnet/standard/issues/1565) depending on whether we need `netstandard2.0` or `netstandard2.1`. How do users discover which one to use? And what if there are dependencies that target `netstandard1.x`?

Reply:
Hmm. I had thought netstandard was around in every core sdk for compatibility.

Reply:
Looks like it is. But it is not in the `ref/` directory, it is under `C:\Program Files\dotnet\shared\Microsoft.NETCore.App\3.1.0`. There appears to be one for every installed SDK version. And this one is an implementation of `netstandard2.1`.

Of course, for SDKs from .NET Core 2.x, the implementation is `netstandard2.0`.

There is also a standalone reference to .NET Standard 2.1 in `C:\Program Files\dotnet\packs\NETStandard.Library.Ref\2.1.0\ref\netstandard2.1`.

Both seem to get me past the `netstandard` error.

Reply:
Okay, after doing it [this way](https://github.com/ikvm-revived/ikvm/issues/45#issuecomment-1141977376) (excluding `-static` and `-classloader:`), it works. It seems that the `-lib:<ikvmtools>/refs` option somehow causes it to search for `netstandard.dll`, but using `-r:<ikvmtools>/refs/*.dll` does not.

Reply:
> 

can I ask how to use it?